Added balloon notifications with tray icon

Necktrox opened the pull request:
## Description

This pull requests aims to add notification functionality in the tray icon area on Windows task bar. It implements an unique class to manage creation, deletion of the tray icon and to show balloon notifications. The balloon notifications support 4 types of icons for the message and it's also possible to silence the notification sound.
## Features
- Tray icon
- Balloon notifications
- 4 notification icons (default, info, warning, error)
- Enable/disable notification sound
- Spam prevention
## Preview screenshots

![Preview for tray notification](http://i.imgur.com/ejWyJ0o.png)
![Preview for modified settings](http://i.imgur.com/F8W0jrV.png)
## Notes

Any suggestions are welcome.

Commit:
Optimize layout for settings tab 'Interface'

Added space for an additional checkbox and cleaned up the GUI mess

Commit:
Added tray icon and createTrayNotification

Adds an invisible tray icon to the task bar (MTA icon will come later).
Adds a Lua function for scripting purposes to create 'balloon
notifications' for the tray icon with 4 different styles (text-only,
info, warning and error).
Future commits will add various options to settings to configure the
tray notifications (disable / always-quiet /
dont-ignore-notification-blocking).

Commit:
Stability improvements for tray icon

Moved functionality into a seperate class CTrayIcon.
Added CTrayIconInterface for CTrayIcon instance in CCore.
Added visible default application icon for task bar.
Removed argument to specify the notification title (hardcoded by
request).
Tray notifications are not feature-complete yet.

Commit:
Fixed optional argument value for createTrayNotification

Fixed code layout

Commit:
Added icon for balloon and checkbox in settings

Added spam protection for balloons (1 notifications each 750ms).
Added CVAR allow_tray_notifications.
Linked mtaicon.ico from Launcher project.

AleksCore commented:
Wow! Nice job!

qaisjp commented on MTA10/mods/shared_logic/lua/CLuaFunctionDefs.Output.cpp:
```diff
@@ -227,7 +227,7 @@ int CLuaFunctionDefs::CreateTrayNotification ( lua_State* luaVM )
     CScriptArgReader argStream ( luaVM );
     argStream.ReadString ( strText );
-    argStream.ReadString ( strType, "noicon" );
+    argStream.ReadString ( strType, "default" );
```
Nice work.

By the way, this needs to be an enum. See the code for setCameraFieldOfView for an example.

Necktrox replied:
Ok, no problem. I was looking at the enum code earlier, but I wasn't sure if that was required.
@qaisjp

Commit:
Refactored CTrayIconType to MTA enum type

Necktrox commented:
Okay, the only problems I had was to find the correct location for the code parts, but it works.
@qaisjp

qaisjp commented:
Nice work!

Please can you explain what the spam prevention is?

Necktrox commented:
I prevent spam by blocking the function for a little period with GetTickCout64. Without this, a script could crash explorer, which happened on my computer.

qaisjp commented:
Okies, just tested your PR.

First thing I notice is that the MTA icon pops up in my taskbar once I have joined a server. Would it be possible for you to make the icon only appear when notifications are sent?

You seem to already have code in there to destroy the tray icon when you leave a server but this doesn't work correctly (the icon does not get destroyed).

Necktrox commented:
I will work on it in the next days, but IIRC there were issues with sending notifications right after creating the tray icon. Furthermore, the notification disappeared after destroying the tray icon, but I am not sure and I can't test it now.

AlexTMjugador commented:
I think that adding a _title_ variable to the tray balloons class would be a nice touch. That way players would be able to know what is the topic of that message more easily, if the script wants to do so. However, it could be abused a bit to confuse players, but the way I see it formatting the title like "Multi Theft Auto: San Andreas — _CUSTOM THING HERE_" would be enough to avoid that.

I look forward to seeing this PR in MTA's trunk!

Necktrox commented:
@AlexTMjugador I would keep it as it is. You could add a title or topic in the body section with enclosed brackets (e.g. "[News] ...") - @Jusonex proposed to remove the ability to change the title to avoid abuse.

AlexTMjugador commented:
I understand, @Necktrox - that is a fair argument. It was just an idea I thought of when first watching this PR, to improve it even more, and I wanted to share it with the author :)

Commit:
Fixed tray icon not showing/hiding properly on connect/disconnect

Commit:
Addendum to last commit

Forgot to add file to commit
https://github.com/multitheftauto/mtasa-blue/pull/35/commits/29efeffe848fd56cec617d963c9c2c423bf5dc29

Necktrox commented:
@qaisjp 
1) It is possible, BUT it won't be possible to show the MTA icon in the message AND the notification will disappear if you hover in the area, where the tray icon should have been actually.

2) Fixed in (both) last commits.

qaisjp commented:
I agree that we should not have the ability to change the title.

Also, what if someone set the message to "Enter windows password to continue.."? Perhaps we should change the title to **"MTA Server says..."**?

Necktrox commented:
It's not difficult to change the title, but I don't really see what this little change should avoid in practice. It won't stop server owners from showing a GUI input with similiar caption. Can you elaborate your argument?

Commit:
Improve script feedback boolean for createTrayNotification

qaisjp commented:
Fair point. I just think that seeing a Windows notification that comes from MTA on a desktop looks more official than a window inside MTA.

Commit:
Changed tray balloon title to avoid misinterpretation

Necktrox commented:
@qaisjp  I changed the title, but I kept the text for the tray icon tooltip.

![Preview](http://i.imgur.com/nWib9j0.png)

ccw808 commented:
My thoughts:
1. When MTA is not minimized, the notifications should be drawn inside MTA. (i.e. with CGraphics functions)
2. Add an option to enable tray notifications when minimized

Necktrox commented:
I agree with the second point, but I am not sure about the first point, because MTA (not a script) itself could use it to show various notifications and most scripters would rather make their own notification system + visuals than use a non-customizable inbuilt one.

**About the second thought**:
A tray notification, as of now, will only show up when your MTA window is not focused (you can't read them when you are in MTA). And there is a checkbox to enable/disable tray notifications (see the screenshots above).

qaisjp commented:
I agree with @necktrox

ccw808 commented:
Ok, that's fine then.

Commit:
Merge remote-tracking branch 'refs/remotes/multitheftauto/master'
